File: android/app/src/main/java/net/wellvo/android/viewmodels/AuthViewModel.kt
package net.wellvo.android.viewmodels

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import io.github.jan.supabase.auth.status.SessionStatus
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import net.wellvo.android.data.models.AppUser
import net.wellvo.android.network.ApiService
import net.wellvo.android.network.AutoJoinResult
import net.wellvo.android.network.WellvoError
import net.wellvo.android.services.AnalyticsService
import net.wellvo.android.services.AuthService
import net.wellvo.android.services.BiometricService
import net.wellvo.android.ui.navigation.AuthState
import net.wellvo.android.util.SecureStorage
import net.wellvo.android.util.Validation
import androidx.compose.runtime.Immutable
import javax.inject.Inject

@Immutable
data class AuthUiState(
    val phoneNumber: String = "",
    val otpCode: String = "",
    val isAwaitingOTP: Boolean = false,
    val isLoading: Boolean = false,
    val isGoogleLoading: Boolean = false,
    val errorMessage: String? = null,
    val email: String = "",
    val password: String = "",
    val displayName: String = "",
    val isSignUp: Boolean = false,
    val showReauthPrompt: Boolean = false,
    val isResettingPassword: Boolean = false,
    val resetPasswordMessage: String? = null,
    val showBiometricSetupPrompt: Boolean = false,
    val biometricLocked: Boolean = false,
    val authLockoutMessage: String? = null,
    val authLockoutSecondsRemaining: Int = 0
)

@HiltViewModel
class AuthViewModel @Inject constructor(
    private val authService: AuthService,
    private val apiService: ApiService,
    private val analyticsService: AnalyticsService,
    val biometricService: BiometricService,
    private val secureStorage: SecureStorage
) : ViewModel() {

    private val _uiState = MutableStateFlow(AuthUiState())
    val uiState: StateFlow<AuthUiState> = _uiState.asStateFlow()

    // Rate limiting stored in encrypted storage
    private var failedAttempts: Int
        get() = secureStorage.loadInt(SecureStorage.AUTH_FAILED_ATTEMPTS)
        set(value) { secureStorage.saveInt(SecureStorage.AUTH_FAILED_ATTEMPTS, value) }
    private var lockoutUntilMs: Long
        get() = secureStorage.loadLong(SecureStorage.AUTH_LOCKOUT_UNTIL)
        set(value) { secureStorage.saveLong(SecureStorage.AUTH_LOCKOUT_UNTIL, value) }
    private var lockoutCountdownJob: Job? = null
    private var otpVerifyAttempts = 0
    private companion object {
        const val MAX_OTP_ATTEMPTS = 5
    }

    private val _authState = MutableStateFlow<AuthState>(AuthState.Loading)
    val authState: StateFlow<AuthState> = _authState.asStateFlow()

    private val _pendingAutoJoin = MutableStateFlow<AutoJoinResult?>(null)
    val pendingAutoJoin: StateFlow<AutoJoinResult?> = _pendingAutoJoin.asStateFlow()

    init {
        observeSessionStatus()
    }

    private fun observeSessionStatus() {
        viewModelScope.launch {
            authService.sessionStatus.collect { status ->
                when (status) {
                    is SessionStatus.Authenticated -> {
                        fetchUserAndAuthenticate()
                    }
                    is SessionStatus.NotAuthenticated -> {
                        val isRefreshFailure = status.isSignOut.not()
                        if (isRefreshFailure && _authState.value is AuthState.Authenticated) {
                            _uiState.value = _uiState.value.copy(showReauthPrompt = true)
                        }
                        _authState.value = AuthState.Unauthenticated
                    }
                    is SessionStatus.Initializing -> {
                        _authState.value = AuthState.Loading
                    }
                    is SessionStatus.RefreshFailure -> {
                        _uiState.value = _uiState.value.copy(showReauthPrompt = true)
                        _authState.value = AuthState.Unauthenticated
                    }
                }
            }
        }
    }

    private suspend fun fetchUserAndAuthenticate() {
        val user = authService.getCurrentUser()
        if (user != null) {
            checkAutoJoin()
            _authState.value = AuthState.Authenticated(user = user)
            checkBiometricSetupPrompt()
        } else {
            val userId = authService.currentUserId()
            if (userId != null) {
                checkAutoJoin()
                val fallbackUser = AppUser(
                    id = userId,
                    displayName = "",
                    role = net.wellvo.android.data.models.UserRole.Owner,
                    timezone = "America/New_York",
                    createdAt = "",
                    updatedAt = ""
                )
                _authState.value = AuthState.Authenticated(user = fallbackUser)
            } else {
                _authState.value = AuthState.Unauthenticated
            }
        }
    }

    private suspend fun checkAutoJoin() {
        try {
            val response = apiService.autoJoin()
            _pendingAutoJoin.value = apiService.checkAutoJoinResult(response)
        } catch (_: Exception) {
            _pendingAutoJoin.value = null
        }
    }

    fun clearAutoJoin() {
        _pendingAutoJoin.value = null
    }

    fun dismissReauthPrompt() {
        _uiState.value = _uiState.value.copy(showReauthPrompt = false)
    }

    fun updatePhoneNumber(phone: String) {
        _uiState.value = _uiState.value.copy(phoneNumber = phone, errorMessage = null)
    }

    fun updateOtpCode(code: String) {
        _uiState.value = _uiState.value.copy(otpCode = code, errorMessage = null)
    }

    fun updateEmail(email: String) {
        _uiState.value = _uiState.value.copy(email = email, errorMessage = null)
    }

    fun updatePassword(password: String) {
        _uiState.value = _uiState.value.copy(password = password, errorMessage = null)
    }

    fun updateDisplayName(name: String) {
        _uiState.value = _uiState.value.copy(displayName = name, errorMessage = null)
    }

    fun toggleSignUp() {
        _uiState.value = _uiState.value.copy(isSignUp = !_uiState.value.isSignUp, errorMessage = null)
    }

    fun sendOTP() {
        val phone = _uiState.value.phoneNumber
        if (!AuthService.isValidUSPhone(phone)) {
            _uiState.value = _uiState.value.copy(errorMessage = "Please enter a valid US phone number.")
            return
        }
        if (isLockedOut()) return

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                authService.sendPhoneOTP(phone)
                otpVerifyAttempts = 0
                _uiState.value = _uiState.value.copy(isLoading = false, isAwaitingOTP = true)
            } catch (e: WellvoError) {
                recordFailedAttempt()
                _uiState.value = _uiState.value.copy(isLoading = false, errorMessage = e.localizedMessage)
            }
        }
    }

    fun verifyOTP() {
        val phone = _uiState.value.phoneNumber
        val code = _uiState.value.otpCode

        if (code.length != 6) {
            _uiState.value = _uiState.value.copy(errorMessage = "Please enter the 6-digit code.")
            return
        }
        if (isLockedOut()) return

        otpVerifyAttempts++
        if (otpVerifyAttempts > MAX_OTP_ATTEMPTS) {
            _uiState.value = _uiState.value.copy(
                errorMessage = "Too many attempts. Please request a new code.",
                isAwaitingOTP = false,
                otpCode = ""
            )
            otpVerifyAttempts = 0
            return
        }

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                authService.verifyPhoneOTP(phone, code)
                analyticsService.track(AnalyticsService.SIGN_IN)
                resetFailedAttempts()
                _uiState.value = _uiState.value.copy(isLoading = false)
            } catch (e: WellvoError) {
                recordFailedAttempt()
                val remaining = MAX_OTP_ATTEMPTS - otpVerifyAttempts
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    errorMessage = "${e.localizedMessage} ($remaining attempts remaining)",
                    otpCode = ""
                )
            }
        }
    }

    fun signInWithEmail() {
        val state = _uiState.value

        if (!Validation.isValidEmail(state.email)) {
            _uiState.value = state.copy(errorMessage = "Please enter a valid email address.")
            return
        }

        if (state.isSignUp) {
            val nameError = Validation.displayNameError(state.displayName)
            if (nameError != null) {
                _uiState.value = state.copy(errorMessage = nameError)
                return
            }
            val pwError = Validation.passwordErrors(state.password)
            if (pwError != null) {
                _uiState.value = state.copy(errorMessage = pwError)
                return
            }
        } else if (state.password.isBlank()) {
            _uiState.value = state.copy(errorMessage = "Please enter your password.")
            return
        }
        if (isLockedOut()) return

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                if (state.isSignUp) {
                    authService.signUpWithEmail(state.email, state.password, state.displayName)
                    analyticsService.track(AnalyticsService.SIGN_UP)
                } else {
                    authService.signInWithEmail(state.email, state.password)
                    analyticsService.track(AnalyticsService.SIGN_IN)
                }
                resetFailedAttempts()
                _uiState.value = _uiState.value.copy(isLoading = false, password = "")
            } catch (e: WellvoError) {
                recordFailedAttempt()
                // Don't reveal if email exists — use generic message for sign-in failures
                val safeMessage = if (state.isSignUp) e.localizedMessage else "Invalid email or password. Please try again."
                _uiState.value = _uiState.value.copy(isLoading = false, errorMessage = safeMessage, password = "")
            }
        }
    }

    fun sendPasswordReset() {
        val state = _uiState.value
        if (!Validation.isValidEmail(state.email)) {
            _uiState.value = state.copy(errorMessage = "Please enter a valid email address.")
            return
        }

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isResettingPassword = true, errorMessage = null, resetPasswordMessage = null)
            try {
                authService.resetPassword(state.email)
            } catch (_: Exception) {
                // Don't reveal whether the email exists
            }
            // Always show same message to avoid user enumeration
            _uiState.value = _uiState.value.copy(
                isResettingPassword = false,
                resetPasswordMessage = "If an account exists with that email, you'll receive a password reset link."
            )
        }
    }

    fun clearResetMessage() {
        _uiState.value = _uiState.value.copy(resetPasswordMessage = null)
    }

    fun signInWithGoogle(context: Context) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isGoogleLoading = true, errorMessage = null)
            try {
                authService.signInWithGoogle(context)
                analyticsService.track(AnalyticsService.SIGN_IN)
                _uiState.value = _uiState.value.copy(isGoogleLoading = false)
            } catch (e: WellvoError) {
                _uiState.value = _uiState.value.copy(isGoogleLoading = false, errorMessage = e.localizedMessage)
            }
        }
    }

    fun signOut() {
        viewModelScope.launch {
            analyticsService.track(AnalyticsService.SIGN_OUT)
            biometricService.reset()
            authService.signOut()
            _uiState.value = AuthUiState()
        }
    }

    // MARK: - Biometric

    fun checkBiometricSetupPrompt() {
        if (biometricService.shouldPromptToEnable()) {
            _uiState.value = _uiState.value.copy(showBiometricSetupPrompt = true)
        }
    }

    fun enableBiometric() {
        biometricService.isEnabled = true
        _uiState.value = _uiState.value.copy(showBiometricSetupPrompt = false)
    }

    fun skipBiometric() {
        biometricService.hasBeenSkipped = true
        _uiState.value = _uiState.value.copy(showBiometricSetupPrompt = false)
    }

    fun setBiometricLocked(locked: Boolean) {
        _uiState.value = _uiState.value.copy(biometricLocked = locked)
    }

    fun backToPhoneEntry() {
        _uiState.value = _uiState.value.copy(isAwaitingOTP = false, otpCode = "", errorMessage = null)
        otpVerifyAttempts = 0
    }

    // MARK: - Rate Limiting

    private fun isLockedOut(): Boolean {
        val until = lockoutUntilMs
        if (until > 0 && System.currentTimeMillis() < until) {
            startLockoutCountdown(until)
            return true
        }
        if (until > 0) {
            lockoutUntilMs = 0
            _uiState.value = _uiState.value.copy(authLockoutMessage = null, authLockoutSecondsRemaining = 0)
        }
        return false
    }

    private fun recordFailedAttempt() {
        failedAttempts += 1
        val count = failedAttempts
        when {
            count >= 10 -> {
                val until = System.currentTimeMillis() + 300_000 // 5 minutes
                lockoutUntilMs = until
                startLockoutCountdown(until)
            }
            count >= 5 -> {
                val until = System.currentTimeMillis() + 30_000 // 30 seconds
                lockoutUntilMs = until
                startLockoutCountdown(until)
            }
        }
    }

    private fun resetFailedAttempts() {
        failedAttempts = 0
        lockoutUntilMs = 0
        otpVerifyAttempts = 0
        lockoutCountdownJob?.cancel()
        _uiState.value = _uiState.value.copy(authLockoutMessage = null, authLockoutSecondsRemaining = 0)
    }

    override fun onCleared() {
        super.onCleared()
        lockoutCountdownJob?.cancel()
        lockoutCountdownJob = null
    }

    private fun startLockoutCountdown(untilMs: Long) {
        lockoutCountdownJob?.cancel()
        lockoutCountdownJob = viewModelScope.launch {
            while (true) {
                val remaining = ((untilMs - System.currentTimeMillis()) / 1000).toInt()
                if (remaining <= 0) {
                    _uiState.value = _uiState.value.copy(authLockoutMessage = null, authLockoutSecondsRemaining = 0)
                    break
                }
                _uiState.value = _uiState.value.copy(
                    authLockoutMessage = "Too many failed attempts. Try again in ${remaining}s.",
                    authLockoutSecondsRemaining = remaining
                )
                delay(1000)
            }
        }
    }
}

File: ios/DailyOK/ViewModels/AuthViewModel.swift
import SwiftUI
import AuthenticationServices
import CryptoKit
import Security
import Supabase

enum AuthState: Equatable {
    case loading
    case unauthenticated
    case authenticated
}

@MainActor
final class AuthViewModel: ObservableObject {
    @Published var authState: AuthState = .loading
    @Published var currentUser: AppUser?
    @Published var errorMessage: String?
    @Published var isLoading = false

    // Sign-up fields
    @Published var email = ""
    @Published var password = ""
    @Published var displayName = ""

    // Phone OTP fields
    @Published var phoneNumber = ""
    @Published var otpCode = ""
    @Published var isAwaitingOTP = false

    // Password reset
    @Published var isResettingPassword = false
    @Published var resetPasswordMessage: String?

    // Biometric
    @Published var showBiometricPrompt = false
    @Published var biometricLocked = false

    // Rate limiting
    @Published var authLockoutMessage: String?
    @Published var authLockoutSecondsRemaining: Int = 0
    private var failedAttempts: Int {
        get { UserDefaults.standard.integer(forKey: "auth_failed_attempts") }
        set { UserDefaults.standard.set(newValue, forKey: "auth_failed_attempts") }
    }
    private var lockoutUntil: Date? {
        get {
            let ts = UserDefaults.standard.double(forKey: "auth_lockout_until")
            return ts > 0 ? Date(timeIntervalSince1970: ts) : nil
        }
        set {
            UserDefaults.standard.set(newValue?.timeIntervalSince1970 ?? 0, forKey: "auth_lockout_until")
        }
    }
    private var lockoutTimer: Task<Void, Never>?
    private var otpVerifyAttempts: Int = 0
    private static let maxOTPAttempts = 5

    /// The raw nonce generated for the current Apple Sign-In attempt.
    /// Stored in Keychain so it survives view recreation and SwiftUI lifecycle events.
    private var currentRawNonce: String? {
        get { KeychainService.load(key: "apple_signin_nonce") }
        set {
            if let newValue {
                _ = KeychainService.save(key: "apple_signin_nonce", value: newValue)
            } else {
                KeychainService.delete(key: "apple_signin_nonce")
            }
            // Migrate: remove old UserDefaults storage
            UserDefaults.standard.removeObject(forKey: "apple_signin_nonce")
        }
    }

    /// Supabase auth state listener handle
    private var authStateTask: Task<Void, Never>?

    init() {
        Task {
            await checkSession()
            listenForAuthStateChanges()
            await checkAppleCredentialRevocation()
            registerForAppleRevocationNotification()
        }
    }

    deinit {
        authStateTask?.cancel()
        NotificationCenter.default.removeObserver(self)
    }

    // MARK: - Session Management

    func checkSession() async {
        do {
            if let user = try await AuthService.shared.currentSession() {
                currentUser = user
                authState = .authenticated
                // Drive the last-seen heartbeat only while signed in.
                HeartbeatService.shared.start()
            } else {
                authState = .unauthenticated
                HeartbeatService.shared.stop()
            }
        } catch {
            authState = .unauthenticated
            HeartbeatService.shared.stop()
        }
    }

    /// Listen for Supabase auth state changes (token refresh, session expiry)
    private func listenForAuthStateChanges() {
        authStateTask = Task {
            for await (event, _) in SupabaseService.shared.client.auth.authStateChanges {
                switch event {
                case .signedIn:
                    if currentUser == nil {
                        await checkSession()
                    }
                case .signedOut, .userDeleted:
                    currentUser = nil
                    authState = .unauthenticated
                    clearFormFields()
                case .tokenRefreshed:
                    // Supabase access tokens last ~1h. Re-mirror the refreshed
                    // token into the shared Keychain so the Notification Service
                    // Extension, widgets, Siri, and watch keep a valid token —
                    // otherwise confirm-delivery / background check-ins start
                    // 401ing after the first refresh (US-IOS084).
                    await SupabaseService.shared.syncAccessTokenToExtension()
                default:
                    break
                }
            }
        }
    }

    // MARK: - Apple Sign-In

    /// Prepare the Apple Sign-In request.
    /// Call this from the SignInWithAppleButton's `onRequest` closure.
    func configureAppleSignInRequest(_ request: ASAuthorizationAppleIDRequest) {
        request.requestedScopes = [.fullName, .email]
        guard let rawNonce = Self.randomNonceString() else {
            // CSPRNG failed: abort this attempt rather than fall back to a
            // non-cryptographic UUID nonce. Leaving currentRawNonce nil makes the
            // result handler reject the credential and ask the user to retry.
            currentRawNonce = nil
            errorMessage = "Couldn't start a secure sign-in. Please try again."
            return
        }
        currentRawNonce = rawNonce
        request.nonce = Self.sha256(rawNonce)
    }

    // MARK: - Nonce Helpers (synchronous, no actor hop needed)

    private static func sha256(_ input: String) -> String {
        let data = Data(input.utf8)
        let hash = SHA256.hash(data: data)
        return hash.compactMap { String(format: "%02x", $0) }.joined()
    }

    /// Returns a cryptographically-random nonce, or `nil` if the system CSPRNG
    /// fails. Callers must treat `nil` as a fatal-for-this-attempt error and
    /// retry — never substitute a non-cryptographic value (a UUID nonce would
    /// weaken the replay protection Apple Sign-In relies on).
    private static func randomNonceString(length: Int = 32) -> String? {
        var randomBytes = [UInt8](repeating: 0, count: length)
        let errorCode = SecRandomCopyBytes(kSecRandomDefault, randomBytes.count, &randomBytes)
        guard errorCode == errSecSuccess else { return nil }
        let charset: [Character] = Array("0123456789ABCDEFGHIJKLMNOPQRSTUVXYZabcdefghijklmnopqrstuvwxyz-._")
        return String(randomBytes.map { charset[Int($0) % charset.count] })
    }

    func signInWithApple(_ result: Result<ASAuthorization, Error>) async {
        isLoading = true
        errorMessage = nil

        switch result {
        case .success(let authorization):
            guard let credential = authorization.credential as? ASAuthorizationAppleIDCredential else {
                errorMessage = "Invalid Apple credential"
                isLoading = false
                return
            }
            guard let rawNonce = currentRawNonce else {
                errorMessage = "Sign-in security check failed. Please try again."
                isLoading = false
                return
            }
            do {
                currentUser = try await AuthService.shared.signInWithApple(credential: credential, rawNonce: rawNonce)
                currentRawNonce = nil
                authState = .authenticated
                clearFormFields()
                await checkBiometricSetupPrompt()
            } catch {
                errorMessage = error.localizedDescription
            }

        case .failure(let error):
            // Don't show error if user cancelled
            if (error as? ASAuthorizationError)?.code != .canceled {
                errorMessage = error.localizedDescription
            }
        }

        isLoading = false
    }

    // MARK: - Link Apple ID

    @Published var hasLinkedApple = false
    @Published var isLinkingApple = false
    @Published var linkAppleMessage: String?

    /// Check whether the current user already has a linked Apple identity.
    func checkAppleLinkStatus() async {
        hasLinkedApple = await AuthService.shared.hasLinkedAppleID()
    }

    /// Configure an Apple Sign-In request for identity linking (reuses nonce logic).
    func configureAppleLinkRequest(_ request: ASAuthorizationAppleIDRequest) {
        request.requestedScopes = [.email]
        guard let rawNonce = Self.randomNonceString() else {
            currentRawNonce = nil
            linkAppleMessage = "Couldn't start a secure sign-in. Please try again."
            return
        }
        currentRawNonce = rawNonce
        request.nonce = Self.sha256(rawNonce)
    }

    /// Handle the Apple Sign-In result for identity linking.
    func linkAppleID(_ result: Result<ASAuthorization, Error>) async {
        isLinkingApple = true
        linkAppleMessage = nil

        switch result {
        case .success(let authorization):
            guard let credential = authorization.credential as? ASAuthorizationAppleIDCredential else {
                linkAppleMessage = "Invalid Apple credential"
                isLinkingApple = false
                return
            }
            guard let rawNonce = currentRawNonce else {
                linkAppleMessage = "Security check failed. Please try again."
                isLinkingApple = false
                return
            }
            do {
                try await AuthService.shared.linkAppleID(credential: credential, rawNonce: rawNonce)
                currentRawNonce = nil
                hasLinkedApple = true
                linkAppleMessage = "Apple ID linked successfully!"
            } catch {
                linkAppleMessage = error.localizedDescription
            }

        case .failure(let error):
            if (error as? ASAuthorizationError)?.code != .canceled {
                linkAppleMessage = error.localizedDescription
            }
        }

        isLinkingApple = false
    }

    // MARK: - Email Auth

    func signInWithEmail() async {
        guard !email.isEmpty, !password.isEmpty else {
            errorMessage = "Please fill in all fields"
            return
        }
        guard !isLockedOut() else { return }

        isLoading = true
        errorMessage = nil

        do {
            currentUser = try await AuthService.shared.signInWithEmail(email: email, password: password)
            authState = .authenticated
            resetFailedAttempts()
            clearFormFields()
            await checkBiometricSetupPrompt()
        } catch {
            errorMessage = error.localizedDescription
            password = "" // Clear password on failure
            recordFailedAttempt()
        }

        isLoading = false
    }

    /// Lightweight RFC-ish email format check so we don't tell the user a reset
    /// link is on the way (or attempt a sign-up) for an obviously-invalid address.
    func isValidEmail(_ value: String) -> Bool {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.range(of: "^[^@\\s]+@[^@\\s]+\\.[^@\\s]+$", options: .regularExpression) != nil
    }

    /// The displayed sign-up password policy, enforced client-side.
    var passwordMeetsPolicy: Bool { PasswordStrength.meetsPolicy(password) }

    func signUpWithEmail() async {
        guard !email.isEmpty, !password.isEmpty, !displayName.isEmpty else {
            errorMessage = "Please fill in all fields"
            return
        }
        guard isValidEmail(email) else {
            errorMessage = "Please enter a valid email address"
            return
        }
        guard passwordMeetsPolicy else {
            errorMessage = "Password must be 10+ characters with an uppercase letter, a lowercase letter, and a number."
            return
        }
        guard !isLockedOut() else { return }

        isLoading = true
        errorMessage = nil

        do {
            currentUser = try await AuthService.shared.signUpWithEmail(
                email: email,
                password: password,
                displayName: displayName
            )
            authState = .authenticated
            resetFailedAttempts()
            clearFormFields()
            await checkBiometricSetupPrompt()
        } catch {
            errorMessage = error.localizedDescription
            password = "" // Clear password on failure
            recordFailedAttempt()
        }

        isLoading = false
    }

    // MARK: - Phone OTP Auth

    func sendPhoneOTP() async {
        let cleaned = phoneNumber.filter(\.isNumber)
        guard cleaned.count >= 10 else {
            errorMessage = "Please enter a valid phone number"
            return
        }
        guard !isLockedOut() else { return }

        isLoading = true
        errorMessage = nil

        do {
            try await AuthService.shared.sendPhoneOTP(phone: phoneNumber)
            isAwaitingOTP = true
            otpVerifyAttempts = 0
        } catch {
            errorMessage = "Could not send verification code. Please try again."
            recordFailedAttempt()
        }

        isLoading = false
    }

    func verifyPhoneOTP() async {
        guard otpCode.count == 6 else {
            errorMessage = "Please enter the 6-digit code"
            return
        }
        guard !isLockedOut() else { return }

        otpVerifyAttempts += 1
        if otpVerifyAttempts > Self.maxOTPAttempts {
            errorMessage = "Too many attempts. Please request a new code."
            isAwaitingOTP = false
            otpCode = ""
            otpVerifyAttempts = 0
            return
        }

        isLoading = true
        errorMessage = nil

        do {
            currentUser = try await AuthService.shared.verifyPhoneOTP(phone: phoneNumber, code: otpCode)
            authState = .authenticated
            resetFailedAttempts()
            clearFormFields()
            await checkBiometricSetupPrompt()
        } catch {
            errorMessage = "Invalid code. Please try again. (\(Self.maxOTPAttempts - otpVerifyAttempts) attempts remaining)"
            otpCode = ""
            recordFailedAttempt()
        }

        isLoading = false
    }

    // MARK: - Password Reset

    func sendPasswordReset() async {
        guard !email.isEmpty else {
            errorMessage = "Please enter your email address"
            return
        }
        guard isValidEmail(email) else {
            // Don't show the "link sent" success for an obviously-wrong address —
            // the user would wait for an email that can never arrive.
            errorMessage = "Please enter a valid email address"
            return
        }

        isResettingPassword = true
        errorMessage = nil
        resetPasswordMessage = nil

        do {
            try await AuthService.shared.resetPassword(email: email)
            // Always show same message to avoid user enumeration
            resetPasswordMessage = "If an account exists with that email, you'll receive a password reset link."
        } catch {
            // Don't reveal whether the email exists
            resetPasswordMessage = "If an account exists with that email, you'll receive a password reset link."
        }

        isResettingPassword = false
    }

    func signOut() async {
        do {
            try await AuthService.shared.signOut()
            await BiometricService.shared.reset()
            HeartbeatService.shared.stop()
            // Drop the shared check-in snapshot so Siri/widget/watch can't act
            // on a stale session after sign-out.
            SharedCheckInPublisher.clear()
            currentUser = nil
            authState = .unauthenticated
            biometricLocked = false
            clearFormFields()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Biometric Authentication

    /// Check if biometric should be presented on app resume.
    func checkBiometricOnResume() async {
        guard authState == .authenticated else { return }
        let biometric = BiometricService.shared
        guard await biometric.isEnabled, await biometric.isBiometricAvailable() else { return }

        biometricLocked = true
        // Withhold the mirrored session from every out-of-process surface while
        // locked, so a widget / Siri / watch tap can't act as the user until
        // biometric auth succeeds. (Belt-and-suspenders with the background
        // withholding in DailyOKApp.)
        SharedCheckInPublisher.withholdTokens()
        await attemptBiometricUnlock()
    }

    /// Run the biometric prompt; on success, lift the lock and restore the
    /// mirrored session for out-of-process surfaces. Called on resume and from
    /// the lock screen's retry button.
    func attemptBiometricUnlock() async {
        let success = await BiometricService.shared.authenticate()
        biometricLocked = !success
        if success {
            await SharedCheckInPublisher.republishTokensFromSession()
        }
    }

    /// After first successful sign-in, check if we should offer biometric setup.
    func checkBiometricSetupPrompt() async {
        guard await BiometricService.shared.shouldPromptToEnable() else { return }
        showBiometricPrompt = true
    }

    func enableBiometric() async {
        await BiometricService.shared.setEnabled(true)
        showBiometricPrompt = false
    }

    func skipBiometric() async {
        await BiometricService.shared.setSkipped(true)
        showBiometricPrompt = false
    }

    // MARK: - Apple Credential Revocation

    /// Check on launch whether the Apple credential has been revoked.
    private func checkAppleCredentialRevocation() async {
        let isValid = await AuthService.shared.checkAppleCredentialStatus()
        if !isValid {
            await signOut()
            errorMessage = "Your Apple ID access was revoked. Please sign in again."
        }
    }

    /// Listen for the system notification that fires when Apple credential is revoked.
    private func registerForAppleRevocationNotification() {
        NotificationCenter.default.addObserver(
            forName: ASAuthorizationAppleIDProvider.credentialRevokedNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            Task { @MainActor in
                await self?.signOut()
                self?.errorMessage = "Your Apple ID access was revoked. Please sign in again."
            }
        }
    }

    // MARK: - Rate Limiting

    /// Check if auth is currently locked out. Returns true if locked.
    private func isLockedOut() -> Bool {
        if let until = lockoutUntil, until > Date() {
            startLockoutCountdown(until: until)
            return true
        }
        // Clear stale lockout
        if lockoutUntil != nil {
            lockoutUntil = nil
            authLockoutMessage = nil
            authLockoutSecondsRemaining = 0
        }
        return false
    }

    /// Record a failed auth attempt and apply lockout if threshold reached.
    private func recordFailedAttempt() {
        failedAttempts += 1
        let count = failedAttempts

        if count >= 10 {
            let lockout = Date().addingTimeInterval(300) // 5 minutes
            lockoutUntil = lockout
            startLockoutCountdown(until: lockout)
        } else if count >= 5 {
            let lockout = Date().addingTimeInterval(30) // 30 seconds
            lockoutUntil = lockout
            startLockoutCountdown(until: lockout)
        }
    }

    /// Reset failed attempts on successful auth.
    private func resetFailedAttempts() {
        failedAttempts = 0
        lockoutUntil = nil
        authLockoutMessage = nil
        authLockoutSecondsRemaining = 0
        otpVerifyAttempts = 0
        lockoutTimer?.cancel()
    }

    private func startLockoutCountdown(until date: Date) {
        lockoutTimer?.cancel()
        lockoutTimer = Task {
            while !Task.isCancelled {
                let remaining = Int(date.timeIntervalSinceNow)
                if remaining <= 0 {
                    authLockoutMessage = nil
                    authLockoutSecondsRemaining = 0
                    break
                }
                authLockoutSecondsRemaining = remaining
                authLockoutMessage = "Too many failed attempts. Try again in \(remaining)s."
                try? await Task.sleep(nanoseconds: 1_000_000_000)
            }
        }
    }

    // MARK: - Helpers

    private func clearFormFields() {
        email = ""
        password = ""
        displayName = ""
        phoneNumber = ""
        otpCode = ""
        isAwaitingOTP = false
        resetPasswordMessage = nil
    }
}
